xarray/core/dataset.py

```diff
@@ -210,6 +224,7 @@ def __init__(self, data_vars=None, coords=None, attrs=None,
             coords = {}
         if data_vars is not None or coords is not None:
             self._set_init_vars_and_dims(data_vars, coords, compat)
+            self._check_multiindex_level_names()
```

Comment:
Ohhh... that makes perfect sense.

In that case, the logic should go into `merge.py`, to ensure these checks will be performed every time variables are modified, too (which doesn't necessarily use the constructor). We should be doing this check on the result of [`merge_variables()`](https://github.com/pydata/xarray/blob/master/xarray/core/merge.py#L97) in `merge_coords_without_align`, `merge_coords` and `merge_core`.